build: record artifact of "Maybe stress{,race}" CI step

tbg opened the pull request:
This previously wasn't the case, but if stressing ran into the test
timeout, TeamCity doesn't provide easy access to the output.

See #43947.

Release note: None

tbg committed:
build: record artifact of "Maybe stress{,race}" CI step

This previously wasn't the case, but if stressing ran into the test
timeout, TeamCity doesn't provide easy access to the output.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43955)
<!-- Reviewable:end -->

andreimatei reviewed (commented):
LGTM

So now there'll be an artifact file with the non-JSON output of the failing test?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43955#-:-LyZuVk98yxwBLwgH_GN:b-8lwqz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)



<!-- Sent from Reviewable.io -->

tbg committed:
cmd/testfilter: improve wrong input format detection

When I wrote testfilter, I wanted to make sure I wasn't accidentally
feeding it non-JSON test output (which would result in zero parsed tests
and thus bogus results). The way I did this was to ensure that there's
at least *some* test JSON in the input.

However, the PR introducing this commit runs in a situation in which
this is too strict, namely that of "maybe stress{,race} pull request".
We will sometimes convert the "empty" input, i.e. the result of invoking
`github-pull-request-make` in situations in which that program
determines not to run any tests.

So we make testfilter smarter: it will only complain about the absence
of test JSON if it also finds something in the input that looks like
test non-JSON.

Release note: None

tbg commented:
Yes, it does (now, I had to fix a bit more stuff to get there):

https://teamcity.cockroachdb.com/repository/download/Cockroach_UnitTests_Testrace/1691827:id/raw.0.json.txt.tgz!/artifacts/full_output.txt

tbg committed:
build: create full human-readable output on test failure

The raw test output is JSON, which isn't pleasant for humans to look at.
There are cases where humans will need to visually inspect the output,
so provide it. (Test output compresses quite well).

Release note: None

tbg commented:
CI failed on known flake https://github.com/cockroachdb/cockroach/issues/43990

bors r=andreimatei
